common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs

```diff
@@ -94,22 +87,26 @@ public class Worker : IWorker, IDisposable {
 
         /// <inheritdoc/>
         public async Task StopAsync() {
-            await _lock.WaitAsync().ConfigureAwait(false);
-            try {
-                if (_cts == null) {
-                    return;
-                }
+            if (_cts == null) {
+                return;
+            }
 
-                _logger.Information("Stopping worker...");
-                _agentConfigProvider.OnConfigUpdated -= ConfigUpdate_Handler;
-                _heartbeatTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+            _logger.Information("Stopping worker...");
+            _agentConfigProvider.OnConfigUpdated -= ConfigUpdate_Handler;
+            _heartbeatTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
 
-                // Inform services, that this worker has stopped working, so orchestrator can reassign job
-                if (_jobProcess != null) {
-                    _jobProcess.Status = WorkerStatus.Stopped;
-                    await SendHeartbeatWithoutResetTimerAsync().ConfigureAwait(false); // need to be send before cancel the CancellationToken
-                }
+            // Inform services, that this worker has stopped working,
+            // so orchestrator can reassign the job
+            if (_jobProcess != null && _jobProcess.Status != WorkerStatus.Stopped) {
+                _jobProcess.Status = WorkerStatus.Stopped;
+                // need to be send before cancel the CancellationToken
+                await SendHeartbeatWithoutResetTimerAsync().ConfigureAwait(false);
+                _jobProcess.Dispose();
+                _jobProcess = null;
```

Comment:
Those 4 lines of code appear here and twice in `ProcessAsync()`. Maybe we can create a dedicated `ResetJobProcessAsync()` method for them ?

Reply:
yes, let's do that when we will decide to do a major refactoring of this stuff.